plugins/uefi/fu-uefi-device.c

```diff
@@ -370,12 +370,17 @@ fu_uefi_device_write_firmware (FuDevice *device, GBytes *fw, GError **error)
 static gboolean
 fu_uefi_device_add_system_checksum (FuDevice *device, GError **error)
 {
-	const gchar *fn = "/sys/class/tpm/tpm0/pcrs";
 	g_autofree gchar *buf = NULL;
 	g_auto(GStrv) lines = NULL;
 
+	/* maybe no TPM */
+	if (!g_file_test ("/sys/class/tpm", G_FILE_TEST_EXISTS)) {
+		g_debug ("no TPM device, skipping");
+		return TRUE;
+	}
+
 	/* get all PCRs */
-	if (!g_file_get_contents (fn, &buf, NULL, error))
+	if (!g_file_get_contents ("/sys/class/tpm/tpm0/pcrs", &buf, NULL, error))
```

Comment:
I tested on an unprovisioned TPM and see `/sys/class/tpm/tpm0` but the `pcrs` file isn't populating.  Possibly because I don't have the ownership claimed.  So might need to special case that too.

Reply:
I didn't test on an unowned TPM. What contents does the `pcrs` file have in that case? a PCR0 of all zeros? Maybe if `/sys/class/tpm/tpm0/owned` = `0` we just return TRUE without looking at the file at all?

Reply:
I'm not even seeing an `owned` file.

```
/sys/class/tpm/tpm0$ ls -alhR
.:
total 0
drwxr-xr-x 4 root root    0 Dec 12 09:32 .
drwxr-xr-x 3 root root    0 Dec 12 09:32 ..
-r--r--r-- 1 root root 4.0K Dec 12 09:32 dev
lrwxrwxrwx 1 root root    0 Dec 12 09:32 device -> ../../../MSFT0101:00
drwxr-xr-x 2 root root    0 Dec 12 09:32 power
drwxr-xr-x 2 root root    0 Dec 12 09:32 ppi
lrwxrwxrwx 1 root root    0 Dec 12 09:32 subsystem -> ../../../../../class/tpm
-rw-r--r-- 1 root root 4.0K Dec 12 09:32 uevent

./power:
total 0
drwxr-xr-x 2 root root    0 Dec 12 09:32 .
drwxr-xr-x 4 root root    0 Dec 12 09:32 ..
-rw-r--r-- 1 root root 4.0K Dec 12 09:32 async
-rw-r--r-- 1 root root 4.0K Dec 12 09:32 autosuspend_delay_ms
-rw-r--r-- 1 root root 4.0K Dec 12 09:32 control
-r--r--r-- 1 root root 4.0K Dec 12 09:32 runtime_active_kids
-r--r--r-- 1 root root 4.0K Dec 12 09:32 runtime_active_time
-r--r--r-- 1 root root 4.0K Dec 12 09:32 runtime_enabled
-r--r--r-- 1 root root 4.0K Dec 12 09:32 runtime_status
-r--r--r-- 1 root root 4.0K Dec 12 09:32 runtime_suspended_time
-r--r--r-- 1 root root 4.0K Dec 12 09:32 runtime_usage

./ppi:
total 0
drwxr-xr-x 2 root root    0 Dec 12 09:32 .
drwxr-xr-x 4 root root    0 Dec 12 09:32 ..
-rw-rw-r-- 1 root root 4.0K Dec 12 09:32 request
-r--r--r-- 1 root root 4.0K Dec 12 09:32 response
-r--r--r-- 1 root root 4.0K Dec 12 09:32 tcg_operations
-r--r--r-- 1 root root 4.0K Dec 12 09:32 transition_action
-r--r--r-- 1 root root 4.0K Dec 12 09:32 version
-r--r--r-- 1 root root 4.0K Dec 12 09:32 vs_operations
```

Reply:
Do you see owned being `1\n` when the device is owned? I guess I just have to look for the file not existing too.

Reply:
Are you working off a TPM1.2 or TPM2.0 with your assumptions?  I just installed the latest TPM2 stack (tpm2-tools, tpm2-tss, tpm2-abrmd) and I do see PCR values even without ownership of the TPM.

```
$ tpm2_pcrlist
sha1 :
  0  : 5c0dec11ab970f5ec8625dfe221f4411beaafa3c
  1  : 2fce29be92f6e655eba85e4698c14e16574c48bc
  2  : b2a83b0ebf2f8374299a5b2bdfc31ea955ad7236
  3  : b2a83b0ebf2f8374299a5b2bdfc31ea955ad7236
  4  : acf30aa210d5265e78aeef172dbdaa7a76a5b2f9
  5  : 4278baa27da3357aebf8a97af2fdcb72c81a77c4
  6  : 178d896f3e7e46828389bc5085a93d48320040e8
  7  : 1feb34e816f434957cd8074e2ab3347a1e1d5ba8
  8  : 0000000000000000000000000000000000000000
  9  : 0000000000000000000000000000000000000000
  10 : 78eac96e18be505a11fe5fdac67fe30d9d75a9a7
  11 : 0000000000000000000000000000000000000000
  12 : 0000000000000000000000000000000000000000
  13 : 0000000000000000000000000000000000000000
  14 : 0108b18d35fe0e8f48a00db263027f32e07dcff6
  15 : 0000000000000000000000000000000000000000
  16 : 0000000000000000000000000000000000000000
  17 : ffffffffffffffffffffffffffffffffffffffff
  18 : ffffffffffffffffffffffffffffffffffffffff
  19 : ffffffffffffffffffffffffffffffffffffffff
  20 : ffffffffffffffffffffffffffffffffffffffff
  21 : ffffffffffffffffffffffffffffffffffffffff
  22 : ffffffffffffffffffffffffffffffffffffffff
  23 : 0000000000000000000000000000000000000000
sha256 :
  0  : 81b8fdf73ea24dc8da1354550e7e6a08f02a3c1aa0e72e5a6cf66e58b60e7c70
  1  : 3201a715111ab6294f5e8f9bb621cb2380ec53586b466b214895e2adaf90f21f
  2  : 3d458cfe55cc03ea1f443f1562beec8df51c75e14a9fcf9a7234a13f198e7969
  3  : 3d458cfe55cc03ea1f443f1562beec8df51c75e14a9fcf9a7234a13f198e7969
  4  : 5611046249e3a32ad3be997d1b2264f4893915637e50a30d8da50c956753d138
  5  : 06ccdfb3881b12047d67bc09b3f594b7adf327e4ea1b769c830c5cac10e0e432
  6  : f0dc4aae3620e35d5ea476a836cc420e51f35e9533eeecc7136e7fc799598bb2
  7  : c7dd67371b93d2f5f4de76c80d219f9b63ee1e92b2e69a0938734cd3db54f803
  8  : 0000000000000000000000000000000000000000000000000000000000000000
  9  : 0000000000000000000000000000000000000000000000000000000000000000
  10 : 2a3375de8a51a4d560954542d41ec7d24216f339aaa7af8519508445d6492cca
  11 : 0000000000000000000000000000000000000000000000000000000000000000
  12 : 0000000000000000000000000000000000000000000000000000000000000000
  13 : 0000000000000000000000000000000000000000000000000000000000000000
  14 : 632959f31641075aa6848d91649123edb324aca48206c605ea0bbe43590dceec
  15 : 0000000000000000000000000000000000000000000000000000000000000000
  16 : 0000000000000000000000000000000000000000000000000000000000000000
  17 : ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
  18 : ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
  19 : ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
  20 : ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
  21 : ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
  22 : ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
  23 : 0000000000000000000000000000000000000000000000000000000000000000
sha384 :
```

Reply:
According to some comments in  https://github.com/stefanberger/swtpm/issues/61 it seems that you can't read pcrs like that for TPM2.0.

Reply:
Do you know if there is a way to get the PCR0 value without shelling out to `tpm2_pcrlist`? ideally I wanted to read the values from sysfs, or use a libtpm2 or something.

Reply:
tpm2-abrmd exports a dbus interface. That's how tpm2_pcrlist communicated to it.

Reply:
New patch WIP.

Reply:
I've added a helper module to support both TPM modes.